src/mlpack/methods/grad_boosting/grad_boosting_impl.hpp

```diff
@@ -0,0 +1,197 @@
+/*
+ * @file methods/grad_boosting/grad_boosting_impl.hpp
+ * @author Abhimanyu Dayal
+ *
+ * Implementation of the Gradient Boosting class.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+
+// Include guard - Used to prevent including double copies of the same code
+#ifndef MLPACK_METHODS_GRAD_BOOSTING_GRADBOOSTING_IMPL_HPP
+#define MLPACK_METHODS_GRAD_BOOSTING_GRADBOOSTING_IMPL_HPP
+
+// Base definition of the GradBoostingModel class.
+#include "grad_boosting.hpp"
+#include <mlpack/core.hpp>
+
+// Defined within the mlpack namespace.
+namespace mlpack {
+
+// Empty constructor.
+template<typename MatType>
+GradBoosting<MatType>::GradBoosting() :
+    numClasses(0),
+    numWeakLearners(0)
+{ /*Nothing to do.*/}
+
+// In case weak learners aren't defined
+/**
+ * Constructor.
+ *
+ * @param data Input data
+ * @param labels Corresponding labels
+ * @param numClasses Number of classes
+ * @param numWeakLearners Number of weak learners
+ */
+template<typename MatType>
+GradBoosting<MatType>::
+  GradBoosting(const MatType& data,
+               const data::DatasetInfo& datasetInfo,
+               const arma::Row<size_t>& labels,
+               const size_t numClasses,
+               const size_t numWeakLearners) :
+  numClasses(numClasses),
+  numWeakLearners(numWeakLearners)
+{
+  TrainInternal(data, datasetInfo, labels, numWeakLearners);
+}
+
+// In case the user inputs the arguments for the Weak Learner
+/**
+ * Constructor.
+ *
+ * @param data Input data
+ * @param labels Corresponding labels
+ * @param numClasses Number of classes
+ * @param numWeakLearners Number of weak learners
+ * @param other Weak Learner, which has been initialized already.
+ */
+template<typename MatType>
+GradBoosting<MatType>::
+GradBoosting(const MatType& data,
+             const data::DatasetInfo& datasetInfo,
+             const arma::Row<size_t>& labels,
+             const size_t numClasses,
+             const size_t numWeakLearners,
+             const size_t minimumLeafSize,
+             const double minimumGainSplit,
+             const size_t maximumDepth) :
+  numClasses(numClasses),
+  numWeakLearners(numWeakLearners)
+{
+  TrainInternal(data, datasetInfo, labels, numWeakLearners,
+                minimumLeafSize, minimumGainSplit, maximumDepth);
+}
+
+template<typename MatType>
+void GradBoosting<MatType>::
+  Train(const MatType& data,
+        const data::DatasetInfo& datasetInfo,
+        const arma::Row<size_t>& labels,
+        const size_t numClasses,
+        const size_t numWeakLearners)
+{
+  SetNumClasses(numClasses);
+  NumWeakLearners(numWeakLearners);
```

Comment:
Why not (a) just pass `numClasses` to `TrainInternal()` and (b) just set `numClasses = numClasses` and `numWeakLearners = numWeakLearners` inside `TrainInternal()` so that you don't even need the function `SetNumClasses()`?  It feels like that would be easier and simpler.